<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2027-1</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Rodrigo Vega Vilchis
### Joel Gómez Uribe
---

## 1.1.3. Estructuras de control, funciones y bibliotecas

En este *Notebook* presentaremos de manera muy breve el uso general de las estrcuturas de control en *Python*, así también, cómo definir y usar funciones dentro de *Python*. Asimismo, mostraremos muy brevemente, cómo leer y escribir archivos (de texto o de datos) y finalmente expondremos cómo crear _**scripts** de Python_.

### Estructuras de control

---

En *Python* existen las estructuras de control comunes usadas en los lenguajes programación, que son las condicionales (`if`) y las iteretivas (`for` y `while`).

Estás se usan junto con, los operadores lógicos y de comparación ya mencionados y que tienen un formato en común en casi todos los lenguajes de programación.

#### Operadores de comparación:

- *Igual que*:  `==`
- *Diferente que*:  `!=`
- *Menor que*:  `<`
- *Mayor que*:  `>`
- *Menor o igual que*:  `<=`
- *Mayor o igual que*:  `>=`

#### Operadores de lógicos:

- *AND*:  `and`
- *OR*:  `or`
- *NOT*:  `!`

A diferencia de otros lenguajes de programación en *Python*, no existen instrucciones que indiquen donde empieza y termina un bloque del programa.

Un bloque empieza por **`:`**, y su extensión se indica por el uso de una cantidad definida de espacio en blanco al principio de cada línea. 

Es muy importante dejar **una identación (un tabulador)** después de que se inicia un bloque de instrucciones definido por `:`.

Algunos editores de texto (como Kate) o IDEs permiten hacer eso de forma automática.

### Estructuras Condicionales

Para checar una condición y ejecutar código dependiente del resultado, se ocupa  **`if`**:

In [1]:
a = int(input('Ingresa el valor de a '))

if (a > 0):
        print("a es positiva")
        a = a + 1
elif a == 0:  # Notese: dos veces =
        print("a es 0")
else:
        print("a es negativa")

Ingresa el valor de a  0


a es 0


In [2]:
a

0

En el primer caso, hay dos comandos en el bloque que se ejecutará en el caso de que `a > 0`. Nótese que `elif` es una abreviación de `else:if`

Las condiciones que se pueden ocupar incluyen: `<`, `<=` y `!=` (*no es igual*). 
Para combinar condiciones se ocupan las palabras `and` y `or`:

In [13]:
a = -3; b = 0

if a > 1 and b > 2:
    print("Grandes")

if a > 0 or b > 0:
    print("Al menos uno positivo")

Aca otro ejemplo

In [22]:
x = int(input("Introduce un número entero no mayor que diez: "))

if x > 10:
    print("Ingresaste un número mayor a diez.")
    print("Déjame arreglar eso por ti.")
    x = 10
    print("tu número es",x)

Introduce un número entero no mayor que diez:  1


Donde podemos usar `and` y `or`:

In [24]:
x = int(input("Introduce un número entero entre 1 y diez: "))
    
if x > 10 and x < 1:
    print("Tu número es demasiado grande o demasiado pequeño.")


Introduce un número entero entre 1 y diez:  0


In [ ]:
x = int(input("Introduce un número entero menor o igual a 1 y menor o igual a diez: "))
    
if x <= 10 and x >=1:
    print("Tu número es correcto.")
else:
    print("Tu número es incorrecto.")

In [ ]:
x = int(input("Introduce un número entero no mayor que diez: "))
if x > 10:
    print("Tu numero es mayor que diez.")
elif x > 9:
    print("Tu número está bien, pero por poco no!")
else:
    print("Tu número está bien. Continua.")

**Ejemplo: Número pares e impares**

Escribamos un programa que tome como entrada un número entero e imprima si el número es par o impar.

In [27]:
n = int(input("Ingresa un entero: "))
if n%2==0:
    print("par")
else:
    print("impar")

Ingresa un entero:  -2


par


## Estructuras Iterativas

La parte central de muchos cálculos científicos consiste en llevar a cabo **ciclos (bucles)**. Para esto usamos las estructura de control **`while`** y **`for`**.

La instrucción `while` ejecuta un bloque de código mientras una cierta condición se satisfaga. Asimismo, la instrucción `for` se usa generalmente cuando se conoce de antemano el número de iteraciones necesarias.

### La estructura `for`

Un ciclo tipo `for` se suele ocupar para llevar a cabo un número de iteraciones que se conoce de antemano. En el bucle de tipo `for`, encontramos una diferencia importante con respecto a lenguajes más tradicionales:

Se puede iterar sobre cualquier lista y ejecutar un bloque de código para cada elemento de la lista:

In [32]:
#r = [ 1, 3, 5]
r = range(0,10)

for n in r:
    print(n)
    print(2*n) #Esto lo aplica para cada n, y una vez termina el ciclo lo reinicia y continúa con el que sigue.
print("Termine!")

0
0
1
2
2
4
3
6
4
8
5
10
6
12
7
14
8
16
9
18
Termine!


In [3]:
l = [1, 2.5, -3.71, "hola", [2, 3]]

for i in l:
    print(2*l) #imprime dos veces la lista
    print(2*i) #imprime dos veces cada elemento, los números los multiplica por 2, a los strings los concatena una vez más
               #A las listas las multiplica dos veces.

[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
2
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
5.0
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
-7.42
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
holahola
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
[2, 3, 2, 3]


In [41]:
type(l)

list

#### La función *range()*

Para averiguar que es lo que hace una función, se puede investigar de manera interactiva:

In [42]:
range?

Init signature: range(self, /, *args, **kwargs)
Docstring:     
range(stop) -> range object
range(start, stop[, step]) -> range object

Return an object that produces a sequence of integers from start (inclusive)
to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
These are exactly the valid indices for a list of 4 elements.
When step is given, it specifies the increment (or decrement).
Type:           type
Subclasses:     

`range(N)` en `python 2.x` devuelve una lista que contiene una progresión aritmética de enteros. Sin embargo en `python 3.x`,`range` es una función que nos devuelve un rango de números naturales a los cuales podemos acceder por medio de un iterador.

In [43]:
range(10)

range(0, 10)

Si queremos iterar sobre muchos elementos, es más útil construir la lista con `range`.

In [44]:
r = range(5)
print("Hola!")

for n in r:
    print("Hola de nuevo!")

Hola!
Hola de nuevo!
Hola de nuevo!
Hola de nuevo!
Hola de nuevo!
Hola de nuevo!


 Por ejemplo, para hacer una iteración para todos los números pares hasta 100, podemos utilizar:

In [45]:
for i in range(100):
    print(2*i)

0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
48
50
52
54
56
58
60
62
64
66
68
70
72
74
76
78
80
82
84
86
88
90
92
94
96
98
100
102
104
106
108
110
112
114
116
118
120
122
124
126
128
130
132
134
136
138
140
142
144
146
148
150
152
154
156
158
160
162
164
166
168
170
172
174
176
178
180
182
184
186
188
190
192
194
196
198


Recordemos que la función `range` crea una lista de `N` elementos comenzado en el *$0$*. O también podemos usar:

In [4]:
for i in range(3,57,3):
    print(2*i)

6
12
18
24
30
36
42
48
54
60
66
72
78
84
90
96
102
108


Para poder indicar **`un inicio`**, **`un fin`** y **`un incremento`** de paso en la iteración.

`range()` también es util para llenar listas:

In [5]:
L = []
print(L)
for i in range(1,10,3):
    L.append(i)
    print(L)

[]
[1]
[1, 4]
[1, 4, 7]


Podemos resumir el código anterior en una sola instrucción.

Por ejemplo, para obtener una _lista_ con los diez primeros números podemos utilizar un `for` (lo explicaremos más adelante) de la siguiente manera:

In [6]:
L1 = [i for i in range(10)]
L1

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Nótese que la lista comienza en 0 (cero) y que **se omite el elemento final**., 

Por ejemplo para obtner una lista de cuatro elementos que llegue hasta el **3** se puede hacer lo siguiente:

In [7]:
[i for i in range(4)]

[0, 1, 2, 3]

También se le puede indicar en donde inicia y donde termina la lista:

In [8]:
[i for i in range(3,10)]

[3, 4, 5, 6, 7, 8, 9]

O bien también se le puede indicar un paso de incremento:

In [9]:
[i for i in range(3,17,2)]

[3, 5, 7, 9, 11, 13, 15]

In [10]:
[i for i in range(3,17,3)]

[3, 6, 9, 12, 15]

In [11]:
[i for i in range(3,17,5)]

[3, 8, 13]

Sobra decir que estas lista generadas por `range()` las podemos almacenar en una variable.

In [12]:
x = [i for i in range(3,17,5)]
print(x)

[3, 8, 13]


### **Ejemplo:** Lineas de emisión de Hidrogeno 

En 1888 Johannes Rydberg publicó su famosa fórmula para las longitudes de onda $\lambda$ de las líneas de emisión del átomo de hidrógeno:

<h3 align="center">$\displaystyle \dfrac{1}{\lambda} = R \left(\dfrac{1}{m^2} - \dfrac{1}{n^2} \right)$</h3> 

donde $R$ es la constante de Rydberg $R = 1.097 × 10^{-2}$ nm$^{-1}$ y $m$ y $n$ son números enteros positivos. Para un valor dado de $m$, las longitudes de onda $\lambda$ dadas por esta fórmula para todo $n > m$ forman una serie, las tres primeras de tales series, para $m = 1$, $2$ y $3$, se conocen como series de *Lyman*, *Balmer* y *Paschen* en honor a sus respectivos descubridores.

Para resumir varias de las cosas vistas, a continuación escribiremos un breve programa de ejemplo, que imprime las longitudes de onda de las primeras cinco líneas en cada una de estas tres series:


In [13]:
R = 1.097e-2

for m in range(1,4):    
    print("\n Serie para m =",m)    
    for n in range(m+1,m+6):
        invlambda = R*(1/m**2-1/n**2)
        print("  ",1/invlambda,"nm")



 Serie para m = 1
   121.5436037678517 nm
   102.55241567912488 nm
   97.23488301428137 nm
   94.95594044363415 nm
   93.76220862091418 nm

 Serie para m = 2
   656.3354603463993 nm
   486.1744150714068 nm
   434.084299170899 nm
   410.2096627164995 nm
   397.04243897498225 nm

 Serie para m = 3
   1875.2441724182836 nm
   1281.9051959890612 nm
   1093.8924339106654 nm
   1005.013673655424 nm
   954.6697605038536 nm


También nótese que `range()` **no acepta** argumentos de punto **flotante**.

In [14]:
[i for i in range(3.,10.)]

TypeError: 'float' object cannot be interpreted as an integer

Recordemos también que `range` sólo funciona con enteros. Por lo que el siguiente programa nos arrojará un error:

In [15]:
p = 10
q = 3
for n in range(p/q):
    print(n)

TypeError: 'float' object cannot be interpreted as an integer

Para poder corregir lo anterior, podemos usar la parte entera:

In [16]:
for n in range(p//q):
    print(n)

0
1
2


O bien la función `arange` de **`numpy`**.

In [17]:
import numpy as np

for n in np.arange(p/q):
    print(n)

0.0
1.0
2.0
3.0


### La estructura `while`


Como instrucción de bucle, la más simple es **`while`**, que como ya mencionamos, ejecuta un bloque de código mientras una cierta condición se satisfaga, y suele ocuparse para llevar a cabo una iteración en la cual no se conoce de antemano el número de iteraciones necesarias. Sin embargo hay que tener cuidado al usar esta estructura pues si no definimos bien la condición de termino puede que el ciclo se repita indefinidamente y nunca acabe. 

Un ejemplo muy sencillo de un while, donde sí se conoce la condición terminal, para contar hasta 9, se puede hacer como sigue:

In [18]:
i = 0
while i < 10:
    # print(i)
    i += 1     # equivalente a i = i + 1
    print(i)

1
2
3
4
5
6
7
8
9
10


O bien, podemos usarlo de la siguiente manera:

In [19]:
x = int(input("Ingresa un número entero no mayor a diez: "))

while x > 10:
    
    print("Este numero es mayor a diez. Vuelve a intentarlo.")
    x = int(input("Ingresa un número entero no mayor a diez: "))

print("Tu número es",x)

Ingresa un número entero no mayor a diez:  99


Este numero es mayor a diez. Vuelve a intentarlo.


Ingresa un número entero no mayor a diez:  88


Este numero es mayor a diez. Vuelve a intentarlo.


Ingresa un número entero no mayor a diez:  787


Este numero es mayor a diez. Vuelve a intentarlo.


Ingresa un número entero no mayor a diez:  8


Tu número es 8


Ahora compiquemos un poco el programa y de los números par e impar y pidamos uno y uno:

In [21]:
print("Ingresa dos enteros, uno par y el otro impar.")

m = int(input("Ingresa el primer entero: "))
n = int(input("Ingresa el segundo entero: "))

while (m+n)%2==0:
    print("Uno debe ser par y el otro impar.")
    
    m = int(input("Ingresa el primer entero: "))
    n = int(input("Ingresa el segundo entero: "))
    
print("Correto! Los números que elegiste son:",m,"y",n)

Ingresa dos enteros, uno par y el otro impar.


Ingresa el primer entero:  8
Ingresa el segundo entero:  2


Uno debe ser par y el otro impar.


Ingresa el primer entero:  1
Ingresa el segundo entero:  2


Correto! Los números que elegiste son: 1 y 2


### **Ejemplo: Sucesión de Fibonacci**

Los números de (la sucesión) de Fibonacci se definen de la siguiente manera:

### $f_n = f_{n-1} + f_{n-2}$

con las condiciones iniciales

#### $f_0=0$  y $f_1=1$

<!-- ![img/fibrabbits.png](attachment:fibrabbits.png) -->
<div>
    <img src="img/fibrabbits.png", width="800" height="600">
</div>

Escribamos un programa para generar los números de Fibonacci, menores a 1000:

In [6]:
fn_1 = 0
fn_2 = 1
while fn_2 < 1000:    
    print(fn_2)
    fn_1,fn_2 = fn_2,fn_1+fn_2

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987


In [36]:
C1 = 0
C2 = 1
n = 18

for i in range (3,n):
    Ci = C1 + C2
    print(Ci)
    C1 = C2
    C2 = Ci

1
2
3
5
8
13
21
34
55
89
144
233
377
610
987


In [47]:
n = 30
C1 = 0
C2 = 1
Cn_2 = C1
Cn_1 = C2
C = []
C.append(C1)
C.append(C2)
for i in range(3,n):
    Cn = Cn_1 + Cn_2 #Itera
    C.append(Cn) #Guarda
    Cn_2 = Cn_1 #Actualiza el siguiente valor de Cn_2
    Cn_1 = Cn #Actualiza el siguiente valor de Cn_1
print(C)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811]


In [48]:
n = 30
C1 = 0
C2 = 1
Cn_2 = C1
Cn_1 = C2
Cn = 1
C = []
C.append(C1)
C.append(C2)
while Cn < 1000:
    Cn = Cn_1 + Cn_2 #Itera
    C.append(Cn) #Guarda
    Cn_2 = Cn_1 #Actualiza el siguiente valor de Cn_2
    Cn_1 = Cn #Actualiza el siguiente valor de Cn_1
print(C)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597]


Una instrucción muy útil, cuando estamos usando un `while` es **`break`**, que nos permiten salir del ciclo incluso si no se cumple la condición. Por ejemplo:

In [ ]:
x = int(input("Ingresa un número entero no mayor a diez: "))

while x>10:

    print("Esto es mayor a diez, por favor intentalo de nuevo.")
    x = x+1
    
    if x==111:
        break

    print(x)


Una instrucción parecida a `break`  es la instrucción **`continue`**. Al colocar esta instrucción dentro `while` (con una condición de un `if`), hará que el programa salte el resto del código del ciclo, regresando al principio del bucle, donde se vuelve a comprobar la condición en el `while`.

**OJO:** Se recomienda un uso bien pensando de esta istrucción, pues podría provocar que nuestro programa se estanque en un *"bucle infinito"*





In [ ]:
x = int(input("Ingresa un número entero no mayor a diez: "))

while x>10:
    
    print("Esto es mayor a diez, por favor intentalo de nuevo.")
    x = x+1
    
    if (x % 2) == 0:
        continue
    elif x==111:
            break
    print(x)

<!-- 
break | continue
- | - 
![break](break.png) | ![continue](continue.png)
 -->

### Diferencia entre el uso de `break`  y `continue`
$\;$

<div align="center">
    <table>      
        <tr>
            <td> <img src="img/break.png" alt="break" style="width: 450px;"/> </td>
            <td> <img src="img/continue.png" alt="continue" style="width: 450px;"/> </td>
        </tr>
        <tr>
            <td><div align="center"><h2>break</h2></div></td>
            <td><div align="center"><h2>continue</h2></div></td>
        </tr>          
    </table>
</div>



### **Ejercicio:** Números de Catalan

Los números de Catalan $C_n$ son una secuencia de números enteros 1, 1, 2, 5, 14, 42, 132 que juegan un papel importante en la mecánica cuántica y la teoría de los sistemas desordenados. (Eran fundamentales para la prueba de Eugene Wigner de la llamada ley del semicírculo). Se definen de la siguiente manera:

<h3 align="center">$C_0 = 1,\qquad C_{n+1} = {4n+2\over n+2}\,C_n$.</h3> 

Escribe un programa que imprima en orden creciente todos los números catalanes menores o iguales a mil millones.


$\text{Pseudocódigo: Es un boceto o esquema de las instrucciones lógicas que le queremos}$ $\text{ ingresar a la computadora, independientemente de la syntaxis del programa.}$

In [28]:
n = 10 #Primero elijo hasta qué número de catalán quiero calcular.
K0 = 1 #Defino mi primer número de catalán y del cual voy a iterar.
Kn = K0 #Quiero que al menos el siguiente número de catalán dependa de la condición inicial.
Catalan = [1] #Solo es para agregar mi condición inicial a la lista.
for i in range(0,n+1): #Para cada número entre 0 y n voy a querer que calcules el n+1-ésimo número de catalán.
    Kn1 = (4*i+2)*Kn/(i+2) #Aquí doy el cómo va a calcular el n+1 número de catalán.
    Kn = Kn1 #Pero lo más importante está aquí, aquí digo que tomes la base del siguiente número de catalán como el previo calculado y así iterarlo.
    Catalan.append(int(Kn1))
print("Los números de catalán ", Catalan)

Los números de catalán  [1, 1, 2, 5, 14, 42, 132, 429, 1430, 4862, 16796, 58786]


In [35]:
K_0 = 1
K_n = K_0
Cat = [1]
K_n1 = 1
while K_n1 < 1000000000: #Aquí le doy la condición al tamaño del número de catalán.    
    K_n1 = (4*i+2)*K_n/(i+2)
    K_n = K_n1
    Cat.append(int(K_n1))
print(Cat)

[1, 3, 12, 42, 150, 525, 1838, 6433, 22518, 78815, 275854, 965491, 3379220, 11827271, 41395451, 144884079, 507094277, 1774829971]


### **Ejercicio:** 

Escribe un programa que realice la siguiente suma:

<h4 align="center">$\displaystyle s = \sum_{k=1}^{100} \dfrac{1}{k}$</h4> 

In [66]:
s = 0
for k in range(1,101):
    s = s + 1/k
print("La suma de los primeros 100 términos es",round(s,3))

La suma de los primeros 100 términos es 5.187


### **Ejercicio:** 

Ahora escribe un programa que realice la suma de los cuadrados de los valores tomados de un archivo de texto:

<h4 align="center">$\displaystyle s = \sum x^2 \; \; \; \forall x \in V_{txt}$</h4> 

In [ ]:
##Pendiente

## Funciones

---
Las funciones se pueden considerar como subprogramas que ejecutan una tarea dada, pueden contener variables propias (locales) así como estructuras de control y cálculos en su interior. En *Python*, las funciones pueden o no aceptar argumentos, y pueden o no regresar resultados.

La sintaxis para declarar una función es como sigue:

In [49]:
def funcion(x):
    print("Argumento x = ", x)
    return x*x

y se llama así:

In [50]:
funcion(3)
funcion(3.5)

Argumento x =  3
Argumento x =  3.5


12.25

Las funciones se pueden utilizar con argumentos de cualquier tipo (el tipo de los argumentos nunca se especifica). Si las operaciones llevadas a cabo no se permiten para el tipo que se provee, entonces *Python* regresa un error:

In [36]:
funcion("hola")

NameError: name 'funcion' is not defined

Las funciones pueden regresar varios resultados al juntarlos en un *tuple* (n-ada):

In [37]:
def f(x, y):
    return 2*x, 3*y

In [38]:
f(3,4)

(6, 12)

Se puede proporcionar una forma sencilla de documentación de una función al proporcionar un *docstring* (cadena de documentación):

In [10]:
def cuad(x):
    """Funcion para elevar un numero al cuadrado.
    Funciona siempre y cuando el tipo de x permita la multiplicacion.
    """
    return x*x

Ahora si interrogamos al objeto `cuad` (con ipython o Jupyter):

In [11]:
cuad?

Signature: cuad(x)
Docstring:
Funcion para elevar un numero al cuadrado.
Funciona siempre y cuando el tipo de x permita la multiplicacion.
File:      c:\users\pc\appdata\local\temp\ipykernel_31344\4072522639.py
Type:      function

nos da esta información. Si la función está definida en un archivo, entonces `cuad??` muestra el código de la definición de la función.

In [12]:
cuad??

Signature: cuad(x)
Source:   
def cuad(x):
    """Funcion para elevar un numero al cuadrado.
    Funciona siempre y cuando el tipo de x permita la multiplicacion.
    """
    return x*x
File:      c:\users\pc\appdata\local\temp\ipykernel_31344\4072522639.py
Type:      function

## BIBLIOTECAS.

---


Se puede escribir un archivo de texto (un *script*) que solo contenga definiciones de funciones, a la cual puede considerarse una biblioteca propia, y también se puede importarse:

`from mi_biblioteca import *`

Así podemos llamar a esta biblioteca desde el *interprete de comandos python* (o el entrono **ipython** o **Jupyter**) y usar sus funciones dentro de este.

O bien, se puede importar detro de otro *script*, con lo que podemos llamar esas funciones o ver sus características coma ya mencionamos.

In [24]:
!touch myBiblio.py

"touch" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.


In [25]:
import miBiblio as mb 

In [26]:
mb.

SyntaxError: invalid syntax (3376479478.py, line 1)

In [27]:
mb.funcion(9)

AttributeError: module 'miBiblio' has no attribute 'funcion'

### **Ejercicio:** 

Supongamos, que tenemos un punto $p$ especificado en coordenadas cilíndricas $r$, $\theta$, $z$. Escribe una función que tome como argumentos los valores de las cordenadas del punto ($r$,$\theta$,$z$) para que regrese la distancia $d$ del origen a $p$ y pruebala para diferentes valores. 

*Hint:* La forma más sencilla es usar el teorema de Pitagoras.

In [40]:
def distancia(r,θ,z):
   return (r**2+z**2)**(1/2)
r = float(input("Ingresa la coordenada radial de tu punto p, r ="))
θ = float(input("Ingresa el ángulo azimutal de tu punto p, θ= "))
z = float(input("Ingresa la altura de tu punto p, z= "))
while r < 0:
    print("No puedes tener distancias radiales negativas, por favor ingresa un valor positivo")
    r = float(input("Ingresa la coordenada radial de tu punto p, r ="))
    θ = float(input("Ingresa el ángulo azimutal de tu punto p, θ = "))
    z = float(input("Ingresa la altura de tu punto p, z = "))

print("La distancia del origen a tu punto p es d = ", round(distancia(r,θ,z),3))


Ingresa la coordenada radial de tu punto p, r = -21
Ingresa el ángulo azimutal de tu punto p, θ=  09
Ingresa la altura de tu punto p, z=  21


No puedes tener distancias radiales negativas, por favor ingresa un valor positivo


Ingresa la coordenada radial de tu punto p, r = 98
Ingresa el ángulo azimutal de tu punto p, θ =  021.3
Ingresa la altura de tu punto p, z =  18


La distancia del origen a tu punto p es d =  99.639


### **Ejercicio:** 

Escribe una función que tome como argumento un número entero y que regrese el factorial del mismo.

<h4 align="center">$\displaystyle n! = \prod_{i=1}^n i$</h4> 

Y pruebala para diferentes valores.

In [66]:
def factorialDiego(n):
    fac = 1
    for i in range(1,n+1):
        fac = fac * i
    return fac

In [67]:
factorialDiego(3)

6

In [74]:
factorialDiego(30)

265252859812191058636308480000000

In [75]:
factorialDiego(6)

720

#### **_Recursividad._**

Una característica muy útil de las funciones que podemos definir es la **_recursividad_**, es decir, la capacidad de una función de llamarse a sí misma. Por ejemplo, considera la siguiente definición del factorial $n!$:

<h4 align="center">
$
      n! = \biggl\lbrace\begin{array}{ll}
        1 & \qquad\mbox{si $n=1$,} \\
        n\times(n-1)! & \qquad\mbox{si $n>1$.}
      \end{array}
$
</h4> 
      
Esto constituye una definición completa del factorial que nos permite calcular el valor de $n!$ para cualquier entero positivo. 


### **Ejercicio:** 

**a)** A partir del ejercicio anterior, convierte (si es que no lo hiciste ya) tu función en una _función recursiva_, es decir, que  cuando la función tome como argumento un número $n$ diferente de $1$, la función se llame a sí misma para calcular el factorial de $n-1$.

In [76]:
def factorial(n):
    if n == 1:
        return 1
    else:
        return n*factorial(n-1)

In [77]:
factorial(1)

1

In [78]:
factorial(7)

5040

**b)** Si hacemos una pequeña modificación a la definición de $C_n$, los números de Catalán se puede reescribir de la siguiente forma:

<h4 align="center">
$
    C_n = \left\lbrace\begin{array}{ll}
      \rule[-9pt]{0pt}{10pt}1 & \qquad\mbox{if $n=0$,} \\
      \dfrac{4n-2}{n+1}\,C_{n-1} & \qquad\mbox{if $n>0$.}
    \end{array}\right.
$
</h4> 

Escribe una función, **usando recursividad**, que calcule $C_n$ y utilízala para calcular e imprimir $C_{100}$.

In [123]:
def catalan(n):
    if n == 0:
        return 1
    elif n > 0:
        return (4*n-2)/(n+1)*catalan(n-1)
cat = [1]
n = int(input("Ingrese hasta qué número de catalán desea calcular, para n ="))
for i in range(1,n):
    cat.append(catalan(i))
Q = map(int,cat)
print(list(Q))

Ingrese hasta qué número de catalán desea calcular, para n = 10


[1, 1, 2, 5, 14, 42, 132, 429, 1430, 4862]


Como puedes ver, el cálculo de los números de Catalán se puede hacer de dos maneras, ya sea directamente o usando _recursividad_.

**c)** Euclides demostró que el máximo común divisor $g(m,n)$ de dos enteros no negativos $m$ y $n$ satisface que:

<h4 align="center">
$
    g(m,n) = \biggl\lbrace\begin{array}{ll}
      m & \qquad\mbox{si $n=0$,} \\
      g(n,m\>\textrm{mod}\>n) & \qquad\mbox{si $n>0$.}
    \end{array}
$
</h4>

Usando esta fórmula escribe una función `g(m,n)` que emplee **recursividad** para calcular el máximo común divisor de $m$ y $n$. Usa tu función para calcular e imprimir el máximo común divisor de 108 y 192.

In [137]:
def g(m,n):
    if n == 0:
        return m
    elif n > 0:
        return g(n, m % n)
x = int(input("Introduce un número entero positivo m ="))
y = int(input("Introduce otro número entero positivo n ="))

while x<0:
    print("Por favor, este primer número tiene que ser positivo.")
    x = int(input("Introdúcelo nuevamente, m ="))
    y = y = int(input("Introduce otro número entero positivo n ="))
if y == 0:
        print("El MCD de ", x, "es él mismo.")
elif y > 0:
        print("El MCD entre", x, "y", y, "es", g(x,y))

Introduce un número entero positivo m = -2
Introduce otro número entero positivo n = -3


Por favor, este primer número tiene que ser positivo.


Introdúcelo nuevamente, m = 3
Introduce otro número entero positivo n = 5


El MCD entre 3 y 5 es 1


In [139]:
def g(m,n):
    if n == 0:
        return m
    elif n > 0:
        return g(n, m % n)
x = int(input("Introduce un número entero positivo m ="))
y = int(input("Introduce otro número entero positivo n ="))

while x<0:
    print("Por favor, este primer número tiene que ser positivo.")
    x = int(input("Introdúcelo nuevamente, m ="))
    y = y = int(input("Introduce otro número entero positivo n ="))
if y == 0:
        print("El MCD de ", x, "es él mismo.")
elif y > 0:
        print("El MCD entre", x, "y", y, "es", g(x,y))

Introduce un número entero positivo m = 102
Introduce otro número entero positivo n = 0


El MCD de  102 es él mismo.


In [140]:
def g(m,n):
    if n == 0:
        return m
    elif n > 0:
        return g(n, m % n)
x = int(input("Introduce un número entero positivo m ="))
y = int(input("Introduce otro número entero positivo n ="))

while x<0:
    print("Por favor, este primer número tiene que ser positivo.")
    x = int(input("Introdúcelo nuevamente, m ="))
    y = y = int(input("Introduce otro número entero positivo n ="))
if y == 0:
        print("El MCD de ", x, "es él mismo.")
elif y > 0:
        print("El MCD entre", x, "y", y, "es", g(x,y))

Introduce un número entero positivo m = 108
Introduce otro número entero positivo n = 192


El MCD entre 108 y 192 es 12


En la mayoría de los casos, si una cantidad se puede calcular **_sin_ recursividad**, será más rápido y normalmente se recomienda hacerlo así, si es posible. Sin embargo, hay algunos cálculos que son esencialmente imposibles (o al menos mucho más difíciles) sin recursividad.

### **Ejercicio:** 

Supongamos que tenemos un número entero $n$ y queremos saber sus **_factores primos_**. Los factores primos se pueden calcular con relativa facilidad dividiendo repetidamente por todos los números enteros desde 2 hasta $n$ y verificando si el resto es cero. 

**a)** Escribe una función que tome como argumento un número $n$ y devuelva una lista de sus factores primos.

In [7]:
def facprimos(n):
    p = []
    
    for i in range(2,n+1):
        while n % i == 0:
            p.append(i)
            n = n//i
        if n == 1:
            break
    return p

z = int(input("¿A qué número entero positivo quieres calcular sus factores primos?"))
while z < 0:
    z = int(input("Debe ser no negativo, ingrésalo de nuevo."))
print("Los factores primos de",z,"son:",facprimos(z))

¿A qué número entero positivo quieres calcular sus factores primos? -3
Debe ser no negativo, ingrésalo de nuevo. 0


Los factores primos de 0 son: []


In [5]:
def facprimos(n):
    p = []
    
    for i in range(2,n+1):
        while n % i == 0:
            p.append(i)
            n = n//i
        if n == 1:
            break
    return p

z = int(input("¿A qué número entero positivo quieres calcular sus factores primos?"))
while z < 0:
    z = int(input("Debe ser no negativo, ingrésalo de nuevo."))
print("Los factores primos de",z,"son:",facprimos(z))

¿A qué número entero positivo quieres calcular sus factores primos? 1290


Los factores primos de 1290 son: [2, 3, 5, 43]


**b)** Usando la función anterior, escribe un programa que imprima los números primos hasta cualquier límite que elijas verificando si tienen un solo factor primo. 

In [25]:
def facprimos(n):
    p = []
    
    for i in range(2,n+1):
        while n % i == 0:
            p.append(i)
            n = n//i
        if n == 1:
            break
    return p

l = int(input("Elije hasta qué número límite deseas verificar a aquellos con un solo factor primo l ="))
primos =[]
for m in range(2, l+1):
    q = facprimos(m)
    if len(q) == 1:
        primos.append(q[0])
print("Los siguientes números son primos", primos)

Elije hasta qué número límite deseas verificar a aquellos con un solo factor primo l = 129


Los siguientes números son primos [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127]


### **Ejercicio:** 

Crea una biblioteca con todas las funciones de los ejercicios de este notebook.

In [26]:
import bibliotecaDiego as bd

In [34]:
bd.catalan??

Signature: bd.catalan(n)
Source:   
def catalan(n):
    """Calcula el n-ésimo número de Catalán"""
    if n == 0:
        return 1
    elif n > 0:
        return (4*n-2)/(n+1)*catalan(n-1)
File:      c:\users\pc\desktop\repositorio_fisica_computacional_2027_1\1_programacionpython\bibliotecadiego.py
Type:      function

In [36]:
bd.distancia??

Signature: bd.distancia(r, θ, z)
Source:   
def distancia(r,θ,z):
    """Calcula la distancia entre un punto en coordenadas cilíndricas P(r,θ,z) al origen."""
    return (r**2+z**2)**(1/2)
File:      c:\users\pc\desktop\repositorio_fisica_computacional_2027_1\1_programacionpython\bibliotecadiego.py
Type:      function

In [37]:
bd.facprimos??


Signature: bd.facprimos(n)
Source:   
def facprimos(n):
    "Descompone un número entero positivo mayor que dos en sus factores primos dentro de una lista."
    p = []
    
    for i in range(2,n+1):
        while n % i == 0:
            p.append(i)
            n = n//i
        if n == 1:
            break
    return p
File:      c:\users\pc\desktop\repositorio_fisica_computacional_2027_1\1_programacionpython\bibliotecadiego.py
Type:      function

In [38]:
bd.factorial??


Signature: bd.factorial(n)
Source:   
def factorial(n):
    """Permite calcular el factorial de un número entero positivo n de forma recursiva."""
    if n == 1:
        return 1
    else:
        return n*factorial(n-1)
File:      c:\users\pc\desktop\repositorio_fisica_computacional_2027_1\1_programacionpython\bibliotecadiego.py
Type:      function

In [39]:
bd.factorialDiego??

Signature: bd.factorialDiego(n)
Source:   
def factorialDiego(n):
    """Calcula el factorial de un número entero positivo n."""
    fac = 1
    for i in range(1,n+1):
        fac = fac * i
    return fac
File:      c:\users\pc\desktop\repositorio_fisica_computacional_2027_1\1_programacionpython\bibliotecadiego.py
Type:      function

In [40]:
bd.mcd??

Signature: bd.mcd(m, n)
Source:   
def mcd(m,n):
    """Encuentra el Máximo Común Divisor entre dos números enteros n y m positivos."""
    if n == 0:
        return m
    elif n > 0:
        return mcd(n, m % n)
File:      c:\users\pc\desktop\repositorio_fisica_computacional_2027_1\1_programacionpython\bibliotecadiego.py
Type:      function